### ISSUE TITLE
`SFTPClient.copy` download and upload the file instead of doing a `'remote copy'`

### ISSUE THREAD

**Issue by khsrali** [ronf/asyncssh#724]
like the title says, steps to reproduce:

```python
import asyncio
import asyncssh
async def reproduce(machine, src_path, dst_path):
    conn = await asyncssh.connect(machine)
    sftp = await conn.start_sftp_client()
    await sftp.copy(src_path, dst_path,)

asyncio.run(reproduce('thor', '/myremote_path/1/100MB.txt', '/myremote_path/100MB.txt') )
```

then when I  monitor my network usage I see 100MB is uploaded and downloaded:
![Screenshot_20241129_123000](https://github.com/user-attachments/assets/45386d7a-fdfa-42ba-a230-797d5914c33a)

**Comment by ronf** [comment]
Are you talking about using the OpenSSH "copy-data" extension? That's not a standard part of the SFTP protocol, and is relatively new even in OpenSSH (only available for OpenSSH 9.0 and later). I can look into supporting this as a new feature on servers capable of it, but whether it is available or not will depend on the server you are connecting to. For all other servers, it will continue to behave as you saw here.

**Comment by khsrali** [comment]
ok, I see. Thanks!

Therefore, now instead of using `SFTPClient.copy`, I will just rely on  `SSHClientConnection.run('cp /src /dest')` .
This would save a lot of traffic for me.

**Comment by ronf** [comment]
Yeah - that should work.

I'll also look into what it might take to support the copy-data extension on systems that support it. For the full-file case in particular, I think this may be straightforward. I'll let you know if I'm able to get that to work.

**Comment by ronf** [comment]
I have a first cut at adding remote copy support. See the patch below. AsyncSSH will automatically use the new `remote_copy()` function when `copy()` or `mcopy()` are called and the source and destination are the same SFTPClient and the server supports this, but you can also call it directly to get full access to all the offset/length options it supports. You just need to open the source and destination files,  pass in those SFTPClientFile objects as arguments, and close them when you're done.

```patch
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index f24de77..fa4a001 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -806,6 +806,22 @@ class _SFTPFileCopier(_SFTPParallelIO[int]):
             if self._progress_handler and self._total_bytes == 0:
                 self._progress_handler(self._srcpath, self._dstpath, 0, 0)

+            if self._srcfs == self._dstfs and \
+                    isinstance(self._srcfs, SFTPClient):
+                try:
+                    await self._srcfs.remote_copy(self._src, self._dst)
+                except SFTPOpUnsupported:
+                    pass
+                else:
+                    self._bytes_copied = self._total_bytes
+
+                    if self._progress_handler:
+                        self._progress_handler(self._srcpath, self._dstpath,
+                                               self._bytes_copied,
+                                               self._total_bytes)
+
+                    return
+
             async for _, datalen in self.iter():
                 if datalen:
                     self._bytes_copied += datalen
@@ -822,8 +838,6 @@ class _SFTPFileCopier(_SFTPParallelIO[int]):
                 setattr(exc, 'offset', self._bytes_copied)

                 raise exc
-
-
         finally:
             if self._src: # pragma: no branch
                 await self._src.close()
@@ -2472,6 +2486,7 @@ class SFTPClientHandler(SFTPHandler):
         self._supports_fsync = False
         self._supports_lsetstat = False
         self._supports_limits = False
+        self._supports_copy_data = False

     @property
     def version(self) -> int:
@@ -2692,6 +2707,8 @@ class SFTPClientHandler(SFTPHandler):
                 self._supports_lsetstat = True
             elif name == b'limits@openssh.com' and data == b'1':
                 self._supports_limits = True
+            elif name == b'copy-data' and data == b'1':
+                self._supports_copy_data = True

         if version == 3:
             # Check if the server has a buggy SYMLINK implementation
@@ -3090,6 +3107,26 @@ class SFTPClientHandler(SFTPHandler):
         else:
             raise SFTPOpUnsupported('fsync not supported')

+    async def copy_data(self, read_from_handle: bytes, read_from_offset: int,
+                        read_from_length: int, write_to_handle: bytes,
+                        write_to_offset: int) -> None:
+        """Make an SFTP copy data request"""
+
+        if self._supports_copy_data:
+            self.logger.debug1('Sending copy-data from handle %s, '
+                               'offset %d, length %d to handle %s, '
+                               'offset %d', read_from_handle.hex(),
+                               read_from_offset, read_from_length,
+                               write_to_handle.hex(), write_to_offset)
+
+            await self._make_request(b'copy-data', String(read_from_handle),
+                                     UInt64(read_from_offset),
+                                     UInt64(read_from_length),
+                                     String(write_to_handle),
+                                     UInt64(write_to_offset))
+        else:
+            raise SFTPOpUnsupported('copy-data not supported')
+
     def exit(self) -> None:
         """Handle a request to close the SFTP session"""

@@ -3142,6 +3179,15 @@ class SFTPClientFile:
         await self.close()
         return False

+    @property
+    def handle(self) -> bytes:
+        """Return handle or raise an error if clsoed"""
+
+        if self._handle is None:
+            raise ValueError('I/O operation on closed file')
+
+        return self._handle
+
     async def _end(self) -> int:
         """Return the offset of the end of the file"""

@@ -4233,6 +4279,35 @@ class SFTPClient:
                                block_size, max_requests, progress_handler,
                                error_handler)

+    async def remote_copy(self, src: SFTPClientFile, dst: SFTPClientFile,
+                          src_offset: int = 0, src_length: int = 0,
+                          dst_offset: int = 0) -> None:
+        """Copy data between remote files
+
+           :param src:
+               The remote file object to read data from
+           :param dst:
+               The remote file object to write data to
+           :param src_offset: (optional)
+               The offset to begin reading data from
+           :param src_length: (optional)
+               The number of bytes to attempt to copy
+           :param dst_offset: (optional)
+               The offset to begin writing data to
+           :type src: :class:`SSHClientFile`
+           :type dst: :class:`SSHClientFile`
+           :type src_offset: `int`
+           :type src_length: `int`
+           :type dst_offset: `int`
+
+           :raises: :exc:`SFTPError` if the server doesn't support this
+                    extension or returns an error
+
+        """
+
+        await self._handler.copy_data(src.handle, src_offset, src_length,
+                                      dst.handle, dst_offset)
+
     async def glob(self, patterns: _SFTPPaths,
                    error_handler: SFTPErrorHandler = None) -> \
             Sequence[BytesOrStr]:
```

I still haven't implemented the AsyncSSH supporting this as a server, and I need to add unit tests, but I thought I'd provide this first cut if you want to try it against an OpenSSH server.

**Comment by khsrali** [comment]
Cheers, very nice!

P.S. ([aiida-core](https://github.com/aiidateam/aiida-core)) is trying to move on from `paramiko` and start using on `asyncssh` to gain more performance on transport tasks.

**Comment by ronf** [comment]
I've finished the server-side support and unit tests for this feature, and it is now available as commit 743966d on the "develop" branch. If you have any problems with it, please let me know!

Also, thanks for letting me know about the AiiDA work. It's always nice hearing about how AsyncSSH is being used. If there's anything I can do to help with your porting effort, reach out any time.

**Comment by khsrali** [comment]
The commit looks very good!
Then we directly use `remote_copy`, if that raises, we'll execute a `cp` command. 

Thanks a lot! 
I think we can close this issue, now as complete

**Comment by ronf** [comment]
FYI, there's one last bit of work I'm planning to do to make the new `remote_copy()` function easier to use. In addition to taking already open `SFTPClientFile` instances, I'm going to allow it to take pathnames in the form of str, bytes, or PurePath like the other get/put/copy functions do. So, in an example like yours, you won't have to worry about explicitly opening/closing the files when calling `remote_copy()`.

I'll let you know here when this change is available.

**Comment by ronf** [comment]
Ok - support for passing in pathnames to `remote_copy()` is now available in commit 4cf2a39 in the "develop" branch.

**Comment by khsrali** [comment]
Thanks a lot, @ronf! 

Now, that I'm trying to deploy it, I face some practical needs. The main difficulty is, one does not know in advance if a server supports remote copy. So one has to call `remote_copy` once, which would only serve the purpose of _check-if-the-server-supports-it_.
Simply because if the server supports it, I'd rather use `copy` & `mcopy` (!) because they support copying directories, glob patterns, etc. which are absent in  `remote_copy` and frankly I would not vote for re-developing them.

if I may take this chance, I have two suggestions:
#### Suggestion A (more desirable)
I would suggests `copy` and `mcopy` be explicitly instruct what to do, with a default behavior . Maybe like:
```python
async def copy(self, ..., strictly_remote = False)                
```

- `strictly_remote = True` : This way, would raise if remote copy is not supported.
- `strictly_remote = False`: Still would do remote copy if the server supports it, otherwise will do download & upload.

This way, I can easily understand whats happning behind the scene and I would only need to handle an error and execute `cp`, for unsupported cases. 

Whereas in implementation as https://github.com/ronf/asyncssh/commit/e72d6421b14ce02ab2610838f2e34bd708ec7804 I first have to do a "fake" `remote_copy` , to see if it raises, if not, then I call on `copy` & `mcopy` (easier for directories, glob, etc.), and yet, I'm paranoid, because I would never get a guarantee that `copy` & `mcopy` will always do a remote copy when supported, because it's not exposed in the api. 

#### Suggestion B
There would be a way to inquiry what sort of things the server that I'm connecting is supporting, so I would know what is happening behind the sense. For example `self._supports_copy_data` could be public and maybe also directly accessible from `SFTPClient`.


P.S. sorry, maybe I shouldn't have close this issue too early, :grimacing: :smile:

**Comment by ronf** [comment]
Thanks for the suggestions. It turned out to be only a few extra lines of code to implement both suggestion A & B here, and having suggestion B to return whether remote copy was available made the code for suggestion A more readable.

I went with a slightly different name here ("remote_only" rather than "strictly_remote"), but the meaning is the same as what you suggested. As you said, this new argument only makes sense in the copy() and mcopy() methods of SFTPClient, as other calls such as get() and put() will never support remote copy.

If you set remote_only and the feature is not available, SFTPOnUnsupported will be raised, similar to what happens if you try to use the new remote_copy() function on a server that doesn't support it.

 In addition, there's a new "supports_remote_copy" property on SFTPClient which returns a boolean indicating if the server supports remote copy. In some cases, it might be cleaner to test this first rather than trying to catch the SFTPOpUnsupported exception.

These changes are available in commit 033ef54 on the "develop" branch.

**Comment by khsrali** [comment]
Wow! that was fast :))
I think now it's perfect, Cheers!

**Comment by ronf** [comment]
t occurred to me overnight that there might be something still not quite right here. You'll probably want to have copy()/mcopy() trigger the error handler when remote_only is set and the remote system doesn't support remote copy. This should only kick in when you are copying a plain file, though, and not when creating a directory or symlink. That shouldn't be too difficult to do, but it means I'll need to move where I do the remote_only check. I'll let you know when I have an update on that.

**Comment by ronf** [comment]
Ok - commit e8169bf should have these changes. In a quick test, I was able to do a recursive wildcard copy with a mix of directories, symlinks, and plain files and an error handler set, and the handler was called once per plain file with the srcpath and dstpath as members of the exception. So, you should be able to use that to trigger a "cp" command remotely while letting AsyncSSH handle all the other copy operations.

**Comment by ronf** [comment]
The name "copy-data" is what the SFTP extension is named on the wire. I stuck with that naming in SFTPClientHandler (and all the other supported SFTP extensions), as that class deals with constructing and parsing the various SFTP request & response messages. SFTPClientHandler exposes a "supports_copy_data" property, but application code shouldn't have direct access to the handler object. The only thing with that access is the SFTPClient class, and that exposes the property "supports_remote_copy" to application code.

**Comment by khsrali** [comment]
I was confused, because I thought they are both withing the same class, now it's clear, thanks for the explanation.

**Comment by ronf** [comment]
This feature is now available in AsyncSSH 2.19.0.
